### PR TITLE
Added Custom Performance Monitor

### LINKED ISSUE #229: Add support for defining custom perfomance monitors
**Describe the project you are working on:** 
Job Queue

**Describe the problem or limitation you are having in your project:** 
Monitoring Performance

**Describe how this feature / enhancement will help you overcome this problem or limitation:** 
It seems to me that Godot has a fantastic monitoring system for debugging. FPS or physics times or memory use or physics interactions, Godot has a monitor for so many of these. But only core engine features. If a Godot user wants to monitor something else, he has to create his own monitoring system (usually just a label that updates every frame or so).

Can we expose this monitoring system to the user? Allow developers to add metrics to the performance monitoring system and display those in the debugger.

For me, I would like to see how many jobs are completed in a given frame and how that is impacting or influencing the frame rate. Later I'd like to monitor how many of any given entity I have in a scene. Or some other property of those entities.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**

```python
#set value every frame
Performance.set_monitor("jobs_completed",jobs_completed)

#as a callback, like set once in `_ready()` and 
#called by the Performance singleton as needed
Performance.create_monitor("jobs_completed",self,"get_jobs_completed")
```
I believe the callback version is "better" because it removes ambiguity of what happens when the function is called more than once in a single frame. Does it add? Does it replace? Does it average them? Doesn't matter, it doesn't happen. But I come from a JS background, so the world is callbacks to me.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

This would be a core editor feature and likely cannot be added without modifying the debugger itself.

**Is there a reason why this should be core and not an add-on in the asset library?:**

I don't think the debugger can be modified this way with the asset library.

### PR DESCRIPTION
**Pull request by simpuid**
Task 1 & 2 of [GSoC project](https://gist.github.com/simpuid/35bd3fa74d052ec1e0bae5453f9fdd12)

Custom monitors can be added/removed/check using `Performance.add_custom_monitor`/`Performance.remove_custom_monitor`/`Performance.has_custom_monitor`

The value can be viewed in the `Custom` section of `Monitor` tab of Debugger.
Basically...
```
func _enter_tree():
	var self_var = self
	Performance.add_custom_monitor("Hello/There", Callable(self_var, "monitor_call"), [1, 2])
	Performance.add_custom_monitor("", Callable(self_var, "monitor_call"), [1, 2])
	Performance.add_custom_monitor("General/Kenobi", Callable(self_var, "monitor_call"), [2, 3])
	Performance.add_custom_monitor("Multiple/First", Callable(self_var, "monitor_call"), [3, 4])
	Performance.add_custom_monitor("Multiple/Second", Callable(self_var, "monitor_call"), [4, 5])
	Performance.add_custom_monitor("Audio/BuiltinCategory", Callable(self_var, "monitor_call"), [5, 6])
	Performance.add_custom_monitor("Too/Much/Slashes/So it's in Custom", Callable(self_var, "monitor_call"), [6, 7])
	Performance.add_custom_monitor("No slash so it's in Custom", Callable(self_var, "monitor_call"), [7, 8])

func _exit_tree():
	Performance.remove_custom_monitor("Hello/There")
	Performance.remove_custom_monitor("")
	Performance.remove_custom_monitor("General/Kenobi")
	Performance.remove_custom_monitor("Multiple/First")
	Performance.remove_custom_monitor("Multiple/Second")
	Performance.remove_custom_monitor("Audio/BuiltinCategory")
	Performance.remove_custom_monitor("Too/Much/Slashes/So it's in Custom")
	Performance.remove_custom_monitor("No slash so it's in Custom")

func monitor_call(from, to) -> float:
	return rand_range(from, to)

```
Above code causes

sample

Edit: Monitor list is now cleared when debugging is started.
Edit: Updated with categorized monitors.
Edit: This implements godotengine/godot-proposals#1014 too

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/229.*

### CODE REVIEW

**Comment by samdze** [comment]
Nice!
Would it be possible to add the capability to define monitors inside any category in the monitors tree? (Not only in the arbitrary "Custom" one)

Like this:
```
func _enter_tree():
	var self_var = self

        // This one goes under "Ammo"
	Performance.add_custom_monitor("Ammo/Gun", ..., ...)
        Performance.add_custom_monitor("Ammo/Rifle", ..., ...)

        // These one go under another category "Enemies"
	Performance.add_custom_monitor("Enemies/Bats", ..., ...)
        Performance.add_custom_monitor("Enemies/Spiders", ..., ...)

        // And you could also add a monitor to an already existing category
	Performance.add_custom_monitor("Physics 3d/Bouncing", ..., ...)

        // The method call could be changed this way if preferred:
        Performance.add_custom_monitor(category, name, callable, params)
```
I think the user would be able to organize his monitors much better.

Maybe I'd also drop the "_custom" from the method names, there's no need to specify those are "custom" monitors that you're adding/removing.

**Comment by simpuid** [comment]
`add_monitor("category/name" , callable...., params....)` seems the best option.

`add_monitor("category", "name", callable..., params....)` will cause problems because internally the ids used for those monitors would be something like `category/name`(That slash could be any ASCII symbol). Method calls like `Performance.add_monitor("a/b", "c", callable..., params...)` would cause confusion among users.

I think something like this will work
* `/` is present => use the supplied category
* `/` is absent => use the `Custom` category.

Adding monitors to built-in categories seems difficult but I think I can do it. However keep in mind that there is no `Physics 3d` . The `Performance` class uses snake_case internally i.e. `physics_3d` is converted to `Physics 3d`, so something like `physics_3d/bouncing` will use the builtin category. This opens the issue of name collisions with monitors like `Physics 3d/bouncing` with `physics_3d/bouncing`
@Faless @mhilbrunner What are your opinions about categorizing custom monitors?

**Comment by simpuid** [comment]
Accidently clicked the `Close and comment`

**Comment by simpuid** [comment]
@samdze  Went ahead with Capitalized category names. (`Physics 3d/something` will work now).
The only limitation is that `Custom/something` and `something` both will be displayed in `Custom` category.

**Comment by samdze** [comment]
@simpuid Seems good!

I also saw that renaming the methods to `add_monitor(...)`, `get_monitor(...)` and `has_monitor(...)` would collide with the already existing `get_monitor(Monitor p_monitor)`, used for built-in monitors.
So while I'd prefer not having that "_custom" in the method names, it's probably necessary if a more extensive refactor is unwanted.

**Comment by samdze** [comment]
I also opened this proposal, you may be interested.
https://github.com/godotengine/godot-proposals/issues/1014

**Review by Faless — COMMENTED** [review]
This looks great! Excellent work :medal_sports: . I'm wondering if we should move the performance drawing functions (setup and parsing) to its own `EditorDebuggerPerformance` class. What do you think @simpuid ?

EDIT: That would also help keeping things clean if we implement godotengine/godot-proposals#1014 in the future (which I think it's a good idea).

**Comment by simpuid** [comment]
> This looks great! Excellent work medal_sports . I'm wondering if we should move the performance drawing functions (setup and parsing) to its own `EditorDebuggerPerformance` class. What do you think @simpuid ?
> 
> EDIT: That would also help keeping things clean if we implement [godotengine/godot-proposals#1014](https://github.com/godotengine/godot-proposals/issues/1014) in the future (which I think it's a good idea).

Yes, we can do that (just like `EditorProfiler`, `EditorVisualProfiler` and `EditorNetworkProfiler`). I can also take help from their implementations.

Shouldn't we name it `EditorPerformance`? (there is no "debugger" word in `EditorProfiler`, `EditorVisualProfiler` and `EditorNetworkProfiler`)

**Comment by Faless** [comment]
> Shouldn't we name it `EditorPerformance` / `EditorPerformanceProfiler`?

Yeah, I agree, both sounds good.
Maybe the latter is more in line with `RemoteDebugger::PerformanceProfiler` in `remote_debugger.cpp`

**Comment by simpuid** [comment]
> Yeah, I agree, both sounds good.
> Maybe the latter is more in line with `RemoteDebugger::PerformanceProfiler` in `remote_debugger.cpp`

@Faless I rewrote the parsing and drawing logic to a separate class `EditorPerformanceProfiler`
The implementation deviates from the proposal but it's better.
Changes:
* Built-in performance monitors' data are stored the same way as custom monitor.
* There is only one drawing function `EditorPerformanceProfiler::_monitor_draw`, godotengine/godot-proposals#1014 could be implemented there for both type of monitors.
* There are only 2 types of messages instead of three
    1. 'performance:profile_frame' : contains the data of all (builtin + custom) monitors.
    2. 'performance:profile_names' : contains the name of custom monitors.

**Comment by simpuid** [comment]
@samdze Added horizontal guide lines and a vertical marker line that can be placed by clicking LMB.
I think that should be enough for godotengine/godot-proposals#1014

Does text of horizontal guide lines looks too cumbersome?

sample

**Comment by Calinou** [comment]
@simpuid There seem to be a lot of horizontal lines on your graph. My PR made it so the number of lines depends on the monitor's height (and never exceeds 5).

**Comment by simpuid** [comment]
> @simpuid There seems to be a lot of horizontal lines on your graph. [My PR](https://github.com/godotengine/godot/pull/39719) made it so the number of lines depends on the monitor's height (and never exceeds 5).

I am calculating the number of horizontal lines depending on the graph height and text height (So they can be rendered easily)
```
int line_count = rect.size.height / (graph_font->get_height() * 2);
```
@Calinou should I limit it to max 5?

**Comment by Calinou** [comment]
> @Calinou should I limit it to max 5?

Yes, there's not much benefit to having more than 5.

**Comment by simpuid** [comment]
Added documentation

**Review by Faless — APPROVED** [review]
Looking good! Amazing job :trophy: !

**Review by Faless — CHANGES_REQUESTED** [review]
Actually, I think you accidentally committed ProjectSetting.xml, could you revert that? Thanks

**Comment by simpuid** [comment]
> Actually, I think you accidentally committed ProjectSetting.xml, could you revert that? Thanks

Sorry! Reverted.

File: doc/classes/Performance.xml [hunk 1]
```diff
@@ -5,12 +5,55 @@
...
+			</return>
+			<argument index="0" name="id" type="StringName">
+			</argument>
+			<argument index="1" name="callable" type="Callable">
+			</argument>
+			<argument index="2" name="arguments" type="Array" default="[  ]">
+			</argument>
+			<description>
+				Adds a custom monitor with name same as id. You can specify the category of monitor using '/' in id. If there are more than one '/' then default category is used. Default category is "Custom".
+				[code]
```
Comment (akien-mga) [inline doc/classes/Performance.xml:27]: This should be [codeblock]. [code] is online code like `this`.

File: editor/debugger/editor_performance_profiler.cpp [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "editor_performance_profiler.h"
+
+#include "editor/editor_scale.h"
+#include "editor/editor_settings.h"
+#include "main/performance.h"
+
+EditorPerformanceProfiler::Monitor::Monitor() {
```
Comment (akien-mga) [inline editor/debugger/editor_performance_profiler.cpp:37]: Initialize members directly when declaring them in the header.

File: editor/debugger/editor_performance_profiler.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+public:
+	void reset();
+	void update_monitors(const Vector<StringName> &p_names);
+	void add_profile_frame(const Vector<float> &p_values);
+	List<float> *get_monitor_data(const StringName &p_name);
+	EditorPerformanceProfiler();
+};
+
+#endif //EDITOR_PERFORMANCE_PROFILER_H
```
Comment (akien-mga) [inline editor/debugger/editor_performance_profiler.h:90]: Add space after `//`.

**Comment by akien-mga** [comment]
Thanks!

**Comment by neikeq** [comment]
Your custom monitors will make a fine addition to our collection.

**Comment by williamd1k0** [comment]
Sorry, but is this portable for 3.2 branch? :P

**Comment by simpuid** [comment]
> Sorry, but is this portable for 3.2 branch? :P

I don't think so! Callables are not available in 3.2 branch :(

**Comment by Faless** [comment]
> Sorry, but is this portable for 3.2 branch?

Not really, the whole debugger was heavily refactored in the master branch, backporting is unlikely.

**Pull request merged**